eclipse-theia/theia
# Add Dirty Diff Peek View

#### What it does

Closes #4544.

Allows the user to browse through dirty diff changes similarly to VS Code:

- `Go to Next Change`/`Go to Previous Change` (`Alt+F5`/`Shift+Alt+F5`) commands move the cursor to the next/previous change.

- `Show Next Change`/`Show Previous Change` (`Alt+F3`/`Shift+Alt+F3`) commands show a peek view with the inline diff of the next/previous change.

The peek view can also be opened by clicking on a dirty diff decoration in the gutter. Clicking on the same decoration again closes the peek view.

Basic actions for browsing through the changes and closing the peek view are provided by the peek view itself in its action bar. Other actions such as `Revert Change` and `Stage Change` can be contributed to the action bar by Theia extensions and VS Code plugins.

The implementation fully supports and has been tested with `@theia/git`:

<img width="2028" alt="Screenshot" src="https://github.com/eclipse-theia/theia/assets/8298491/5a0b7fde-e890-4c3f-9046-02bf7b45c5fe">

and `vscode.git`:

<img width="2024" alt="Screenshot" src="https://github.com/eclipse-theia/theia/assets/8298491/a6da007a-87fd-42d7-b167-7f13469d3848">

#### How to test

Verify that the following steps work correctly regardless of whether `@theia/git` or `vscode.git` is used.

0. Open a file with unstaged changes. The editor should be focused.

1. Move the cursor to the beginning of the next/previous change by using `Go to Next Change`/`Go to Previous Change` (`Alt+F5`/`Shift+Alt+F5`) commands.

2. Browse through the changes by using `Show Next Change`/`Show Previous Change` (`Alt+F3`/`Shift+Alt+F3`) commands. Press `ESC` to close the peek view.

3. Click on a dirty diff decoration in the gutter to show the corresponding change. Click on the same decoration again to close the peek view.

4. Open the peek view again and verify that each of the actions in its action bar (browsing through the changes, closing the peek view, reverting the change, staging the change) works as expected.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
